### PR TITLE
fix #13790;  ptr char (+friends) should not implicitly convert to cstring

### LINKED ISSUE #13790: `ptr char` implicitly converts to cstring, resulting in undefined behavior
* `ptr char` implicitly converts to cstring, resulting in undefined behavior
* ditto with `ptr array[N, char]` and `ptr UncheckedArray[N, char]`
this introduces a weird special case in the language (eg `$a` doesn't compile for other types a: ptr[T] except for T=char) as well as weird bugs and undefined behavior in seemingly safe code

### Example 1
```nim
when true:
  import std/unittest
  var x: ptr char
  var s: cstring
  doAssert cstring isnot ptr char
  doAssert (ptr char) isnot cstring

  proc fun(a: cstring)=discard
  proc fun2(a: ptr char)=discard
  doAssert not compiles(fun2 cstring"asdf")
  doAssert not compiles(block: x = s)

  # BUG
  check not compiles(fun(x))
  check not compiles(echo(x))
  check not compiles($(x))
  check not compiles(block: s = x)
```

### Current Output
```
/Users/timothee/git_clone/nim/timn/tests/nim/all/t10433.nim(15, 8): Check failed: not compiles(fun(x))
compiles(fun(x)) was true
/Users/timothee/git_clone/nim/timn/tests/nim/all/t10433.nim(16, 8): Check failed: not compiles(echo(x))
compiles(echo(x)) was true
/Users/timothee/git_clone/nim/timn/tests/nim/all/t10433.nim(17, 8): Check failed: not compiles($(x))
compiles($(x)) was true
```

### Expected Output
tests pass
ie, it should be illegal to do those implicit conversions from ptr char to cstring

### Example 2
in case you may think it's a good idea to have implicit conversion from ptr char to cstring, consider this:
```nim
when true:
  proc fun(a, b: var openArray[char]) =
    echo a[0].addr # prints abcdef sometimes, or sometimes other garbage
  proc main()=
    var a = ['a','b', 'c']
    var b = ['d','e', 'f']
    fun(a,b)
  main()
```

you get (different) garbage results when recompiling the same program:
```
nim c -r $timn_D/tests/nim/all/t10433.nim
abc`UG
nim c -r $timn_D/tests/nim/all/t10433.nim
abc`%
```

### Example 3
in other cases, you'll end up with buffer overflow bugs/attacks, since a `ptr char` could come from any source (eg ffi calls, malloc / `createU`, etc) with nothing that should guarantee that it's 0 terminated.

That problem is exacerbated when a `ptr char` field is nested somewhere deep inside an object and we are calling `$` on it

### Example 4
likewise, implicit conversion from `ptr array[0..2, char]` to `cstring` should be disallowed. It has the same caveats, eg:
```nim
when true:
  var vals = @[['a','b','c'], ['d','e','f']]
  echo vals[0]
  echo vals[0].addr # should not compile (unless `$` has an overload for ptr)
  var z: cstring = vals[0].addr # should not compile
```
prints
```
['a', 'b', 'c']
abcdef # error prone!
```

and ditto for ptr UncheckedArray, eg:
```
when true:
  var vals1 = @[['a','b','c'], ['d','e','f']]
  echo cast[ptr UncheckedArray[char]](vals1[0].addr) # prints
```

### Possible Solution
* prevent implicit conversion (sigmatch) from X to cstring where X is not cstring (eg `ptr char`, `ptr[array[N, char]` etc)
* add an easy way to convert explicitly, eg:
```nim
template toCstring*(a: ptr char): cstring = cast[cstring](a)
```

* if the worry is about backward compatibility, then make it an error but allow a flag:
`--legacy:implicitPtrCharToCstring` (adding to existing list `--legacy:allowSemcheckedAstModification|checkUnsignedConversions`)

### Additional Information
* latest devel 0a49fe5fad36c917f459420d832cda306d449d6e 1.1.1
* since at least 0.17 or forever
* D20200327T225837

### PR DESCRIPTION
**Pull request by ringabout**
fix #13790
ref https://github.com/timotheecour/Nim/commit/b9ecff1f0c2fbd99ff1879c8cc08999940788995

Explicit conversion is disabled in this PR too, since the conversion is unsafe.

It breaks at compile time.

### CODE REVIEW

**Comment by arnetheduck** [comment]
might be a good idea to add a warning + a strict mode instead

File: compiler/sigmatch.nim [hunk 1]
```diff
@@ -1392,15 +1392,18 @@ proc typeRel(c: var TCandidate, f, aOrig: PType,
...
-      if a.len == 1:
-        let pointsTo = a[0].skipTypes(abstractInst)
-        if pointsTo.kind == tyChar: result = isConvertible
-        elif pointsTo.kind == tyUncheckedArray and pointsTo[0].kind == tyChar:
-          result = isConvertible
-        elif pointsTo.kind == tyArray and firstOrd(nil, pointsTo[0]) == 0 and
-            skipTypes(pointsTo[0], {tyRange}).kind in {tyInt..tyInt64} and
-            pointsTo[1].kind == tyChar:
-          result = isConvertible
+      when false:
```
Comment (ringabout) [inline compiler/sigmatch.nim]:
```suggestion
      when true:
```

**Comment by ringabout** [comment]
> might be a good idea to add a warning + a strict mode instead

I agree I will hide it behind `nimPreviewSlimSystem` and gives a warning in lenient mode.

**Comment by arnetheduck** [comment]
> nimPreviewSlimSystem

is the slim system scheduled to become enabled by default?

**Comment by ringabout** [comment]
I'm not sure whether it has been decided that `nimPreviewSlimSystem` will be enabled by default for v2. 

It has been enabled for compiler, `tests/stdlib`, and documentation builds (not tests). If it is going to be enabled for v2, testament will be patched to make important packages' tests green.

**Comment by Araq** [comment]
> is the slim system scheduled to become enabled by default?

No. v2 is mostly about finally making mm:orc the default plus a couple of other things that we consider production ready enough for v2. These are features that exist at least in 1.6 already.

There are new switches like slim system and "strictDefs" and "strictCaseObjects" which have seen their introduction with v2 and thus are feature previews.

**Comment by arnetheduck** [comment]
in that case, I believe this fix should go under a separate flag + that the warning should be available on 1.6 too (so that people can find the implicit conversions in their code starting now) - with the warning in place, it becomes trivial to fix before this becomes enforced

**Comment by arnetheduck** [comment]
also, it would be nice to be able to set the "explicitness" of any compiler message (ie silent/hint/warning/error), so that we can do `--msg[PtrToCstringConv]=warning` while working on it then `--msg[PtrToCstringConv]=error` when we've fixed all existing code and want to put a hard block on new code re-introducing it

**Comment by ringabout** [comment]
done => https://github.com/nim-lang/Nim/pull/20814

**Comment by ringabout** [comment]
`cast[cstring]` refactor will be done later. I will examine some cases first.

**Comment by ringabout** [comment]
> in that case, I believe this fix should go under a separate flag + that the warning should be available on 1.6 too

Done; I think this PR should be backported too.

**Comment by ringabout** [comment]
Compiler and stdlib fixes should be backported at least.

**Comment by metagn** [comment]
This did not account for [`cstring` to `pointer` conversion](https://github.com/nim-lang/Nim/blob/8c55e2999b27e3f088fca2e61739b33498fb07ef/compiler/sigmatch.nim#L1388)

**Pull request merged**